File: J9Ch/src/S_7_Follow_Up/Required_14/python_cpp/_403_Continuous_Subarray_Sum_II_Medium.cpp

class Solution {
public:
    /**
     * @param A an integer array
     * @return  A list of integers includes the index of
     *          the first number and the index of the last number
     */
    vector<int> continuousSubarraySumII(vector<int>& A) {
        // Write your code here
        vector<int> result;
        result.push_back(0);
        result.push_back(0);

        int total = 0;
        int len = A.size();
        int start = 0, end = 0;
        int sum = 0;
        int ans = -0x7fffffff;
        for (int i = 0; i < len; ++i) {
            total += A[i];
            if (sum < 0) {
                sum = A[i];
                start = end = i;
            } else {
                sum += A[i];
                end = i;
            }
            if (sum >= ans) {
                ans = sum;
                result[0] = start;
                result[1] = end;
            }
        }
        sum = 0;
        start = 0, end = -1;
        for (int i = 0; i < len; ++i) {
            if (sum > 0) {
                sum = A[i];
                start = end = i;
            } else {
                sum += A[i];
                end = i;
            }
            if (total - sum > ans && !(start == 0 && end == len-1)) {
                ans = total - sum;
                result[0] = (end + 1) % len;
                result[1] = (start - 1 + len) % len;
            }
        }
        return result;
    }
};

File: J9Ch/src/S_7_Follow_Up/Required_14/python_cpp/_403_Continuous_Subarray_Sum_II_Medium.py


class Solution:
    # @param {int[]} A an integer array
    # @return {int[]}  A list of integers includes the index of the
    #                  first number and the index of the last number
    def continuousSubarraySumII(self, A):
        # Write your code here
        ans = -0x7fffffff
        sum, total = 0, 0
        start, end = 0, -1
        result = [-1, -1]
        length = len(A)
        for x in A:
            total += x
            if sum < 0:
                sum = x
                start = end + 1
                end = start
            else:
                sum += x
                end += 1
            if sum > ans:
                ans = sum
                result = [start, end]


        start, end = 0, -1
        sum = 0
        for x in A:
            if sum > 0:
                sum = x
                start = end + 1
                end = start
            else:
                sum += x
                end += 1
            if start == 0 and end == length-1:
                continue
            if total - sum > ans:
                ans = total - sum
                result = [(end + 1) % length, (start - 1 + length) % length]

        return result